Para encriptar un mensaje, el primer paso es convertirlo en una secuencia de números como lo indica la siguiente tabla:
|A|B|C|D|E|F|G|H|I|J|K|L|M|N|Ñ|O|P|Q|R|S|T|U|V|W|X|Y|Z|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31|32|33|34|35|36|37|

y al espacio entre palabras le asignamos los dígitos $38$. Así, "hola mundo" se convierte en 18262211382332241426.

|H|O|L|A| |M|U|N|D|O|
|---|---|---|---|---|---|---|---|---|---|
|18|26|22|11|38|23|32|24|14|26|

In [1]:
import sympy.ntheory as nt
import sympy as sy
from tools_blocks2message import *
from tools_message2blocks import *
from tools_for_RSA import *

`mensaje1 = "los cientificos se esfuerzan por hacer posible lo imposible mientras que los politicos por hacer lo posible imposible"`

Tomando `p = nt.prime(1000)`,  `q = nt.prime(2000)` y `e = 96558349`. Encripte `mensaje1` usando RSA.

In [2]:
p = nt.prime(1000)
q = nt.prime(2000)
phi_m = (p-1) * (q-1)
e = random_coprime_with(phi_m)
e

8220617

In [3]:
mensaje1 = "los cientificos se esfuerzan por hacer posible lo imposible mientras que los politicos por hacer lo posible imposible"
mis_bloques1 = message2blocks(mensaje1, possible_len=2)

In [4]:
d = inverse_of(e, phi_m)
d

103314329

In [5]:
m = p*q
mis_bloques_encriptados = encode_rsa(m, e, mis_bloques1)
# mis_bloques_encriptados

In [6]:
mis_bloques_desencriptados = decode_rsa(m, d, mis_bloques_encriptados)

In [7]:
mis_bloques1[:15]

[22, 26, 30, 38, 13, 19, 15, 24, 31, 19, 16, 19, 13, 26, 30]

In [8]:
blocks2message(mis_bloques_desencriptados[:])

'los cientificos se esfuerzan por hacer posible lo imposible mientras que los politicos por hacer lo posible imposible'

In [9]:
print("p: ", p)
print("q: ", q)
print("e: ", e)
print(mis_bloques_encriptados)


p:  7919
q:  17389
e:  8220617
[33113290, 80365780, 16594511, 47171377, 119770509, 54937779, 37098340, 125119529, 16827617, 54937779, 47297777, 54937779, 119770509, 80365780, 16594511, 47171377, 16594511, 37098340, 47171377, 37098340, 16594511, 47297777, 18785197, 37098340, 121728071, 73500723, 121944756, 125119529, 47171377, 101452928, 80365780, 121728071, 47171377, 96535797, 121944756, 119770509, 37098340, 121728071, 47171377, 101452928, 80365780, 16594511, 54937779, 19799694, 33113290, 37098340, 47171377, 33113290, 80365780, 47171377, 54937779, 135601836, 101452928, 80365780, 16594511, 54937779, 19799694, 33113290, 37098340, 47171377, 135601836, 54937779, 37098340, 125119529, 16827617, 121728071, 121944756, 16594511, 47171377, 33776723, 18785197, 37098340, 47171377, 33113290, 80365780, 16594511, 47171377, 101452928, 80365780, 33113290, 54937779, 16827617, 54937779, 119770509, 80365780, 16594511, 47171377, 101452928, 80365780, 121728071, 47171377, 96535797, 121944756, 119770509, 3709

1) ¿Cuál es el mensaje oculto en la siguiente lista de enteros, si se sabe que se ha usado el RSA con p=7919, q=17389 y e=96558349?

[116427824, 71262391, 95967677, 108216637, 130284225, 44866252, 16356886, 121976595, 16896540, 44866252, 106409323, 44866252, 130284225, 71262391, 95967677, 108216637, 95967677, 16356886, 108216637, 16356886, 95967677, 106409323, 162951, 16356886, 122552340, 13783210, 87965332, 121976595, 108216637, 86661090, 71262391, 122552340, 108216637, 31583046, 87965332, 130284225, 16356886, 122552340, 108216637, 86661090, 71262391, 95967677, 44866252, 105050342, 116427824, 16356886, 108216637, 116427824, 71262391, 108216637, 44866252, 36622724, 86661090, 71262391, 95967677, 44866252, 105050342, 116427824, 16356886, 108216637, 36622724, 44866252, 16356886, 121976595, 16896540, 122552340, 87965332, 95967677, 108216637, 12960247, 162951, 16356886, 108216637, 116427824, 71262391, 95967677, 108216637, 86661090, 71262391, 116427824, 44866252, 16896540, 44866252, 130284225, 71262391, 95967677, 108216637, 86661090, 71262391, 122552340, 108216637, 31583046, 87965332, 130284225, 16356886, 122552340, 108216637, 116427824, 71262391, 108216637, 86661090, 71262391, 95967677, 44866252, 105050342, 116427824, 16356886, 108216637, 44866252, 36622724, 86661090, 71262391, 95967677, 44866252, 105050342, 116427824, 16356886]

In [10]:
mensaje2 = "gran parte de las dificultades por las que atraviesa el mundo se deben a que los ignorantes estan completamente seguros y los inteligentes llenos de dudas"
mis_bloques2 = message2blocks(mensaje2, possible_len=2)
mis_bloques2_encriptados = encode_rsa(m, e, mis_bloques2)
print(mis_bloques2_encriptados)

[112337747, 121728071, 121944756, 125119529, 47171377, 101452928, 121944756, 121728071, 16827617, 37098340, 47171377, 13773263, 37098340, 47171377, 33113290, 121944756, 16594511, 47171377, 13773263, 54937779, 47297777, 54937779, 119770509, 18785197, 33113290, 16827617, 121944756, 13773263, 37098340, 16594511, 47171377, 101452928, 80365780, 121728071, 47171377, 33113290, 121944756, 16594511, 47171377, 33776723, 18785197, 37098340, 47171377, 121944756, 16827617, 121728071, 121944756, 133461719, 54937779, 37098340, 16594511, 121944756, 47171377, 37098340, 33113290, 47171377, 135601836, 18785197, 125119529, 13773263, 80365780, 47171377, 16594511, 37098340, 47171377, 13773263, 37098340, 19799694, 37098340, 125119529, 47171377, 121944756, 47171377, 33776723, 18785197, 37098340, 47171377, 33113290, 80365780, 16594511, 47171377, 54937779, 112337747, 125119529, 80365780, 121728071, 121944756, 125119529, 16827617, 37098340, 16594511, 47171377, 37098340, 16594511, 16827617, 121944756, 125119529, 

In [11]:
blocks2message(decode_rsa(m, d, mis_bloques2_encriptados))

'gran parte de las dificultades por las que atraviesa el mundo se deben a que los ignorantes estan completamente seguros y los inteligentes llenos de dudas'

2) ¿Cuál es el mensaje oculto en la siguiente lista de enteros, si se sabe que se ha usado el RSA con p=7919, q=17389 y e=96558349?

[94132717, 122552340, 87965332, 121976595, 108216637, 86661090, 87965332, 122552340, 16896540, 16356886, 108216637, 96593042, 16356886, 108216637, 116427824, 87965332, 95967677, 108216637, 96593042, 44866252, 106409323, 44866252, 130284225, 162951, 116427824, 16896540, 87965332, 96593042, 16356886, 95967677, 108216637, 86661090, 71262391, 122552340, 108216637, 116427824, 87965332, 95967677, 108216637, 12960247, 162951, 16356886, 108216637, 87965332, 16896540, 122552340, 87965332, 13625048, 44866252, 16356886, 95967677, 87965332, 108216637, 16356886, 116427824, 108216637, 36622724, 162951, 121976595, 96593042, 71262391, 108216637, 95967677, 16356886, 108216637, 96593042, 16356886, 105050342, 16356886, 121976595, 108216637, 87965332, 108216637, 12960247, 162951, 16356886, 108216637, 116427824, 71262391, 95967677, 108216637, 44866252, 94132717, 121976595, 71262391, 122552340, 87965332, 121976595, 16896540, 16356886, 95967677, 108216637, 16356886, 95967677, 16896540, 87965332, 121976595, 108216637, 130284225, 71262391, 36622724, 86661090, 116427824, 16356886, 16896540, 87965332, 36622724, 16356886, 121976595, 16896540, 16356886, 108216637, 95967677, 16356886, 94132717, 162951, 122552340, 71262391, 95967677, 108216637, 11752584, 108216637, 116427824, 71262391, 95967677, 108216637, 44866252, 121976595, 16896540, 16356886, 116427824, 44866252, 94132717, 16356886, 121976595, 16896540, 16356886, 95967677, 108216637, 116427824, 116427824, 16356886, 121976595, 71262391, 95967677, 108216637, 96593042, 16356886, 108216637, 96593042, 162951, 96593042, 87965332, 95967677]

3. Usando RSA con p=7919, q=17389 y e=96558349, encripte el siguiente mensaje:

`"las matematicas pueden ser definidas como aquel tema del cual no sabemos nunca lo que decimos ni si lo que decimos es verdadero"`

4. Usando RSA con p=7919, q=17389 y e=96558349, encripte el siguiente mensaje:

`"lo mas dificil de aprender en la vida es que puente hay que cruzar y que puente hay que quemar"`

In [12]:
p = nt.prime(10000)
q = nt.prime(20000)
m = p*q
phi_m = (p-1)*(q-1)

In [13]:
e = random_coprime_with(phi_m)
e

10676693633

In [14]:
print("p: ", p)
print("q: ", q)
print("e: ", e)


p:  104729
q:  224737
e:  10676693633


In [15]:
mensaje3 = "me opongo a toda supersticion sea musulmana sea cristiana sea judia o sea budista"
mis_bloques3 = message2blocks(mensaje3, possible_len=8)
mis_bloques3

mensaje4 = "no creas conveniente actuar ocultando pruebas pues las pruebas terminan por salir a la luz a vista de todos"
mis_bloques4 = message2blocks(mensaje4, possible_len=8)
mis_bloques4

[24263813,
 29151130,
 38132624,
 33152419,
 15243115,
 38111331,
 32112938,
 26133222,
 31112414,
 26382729,
 32151211,
 30382732,
 15303822,
 11303827,
 29321512,
 11303831,
 15292319,
 24112438,
 27262938,
 30112219,
 29381138,
 22113822,
 32373811,
 38331930,
 31113814,
 15383126]

In [16]:
mis_bloques3_encriptados = encode_rsa(m, e, mis_bloques3)
# print(mis_bloques3_encriptados)

mis_bloques4_encriptados = encode_rsa(m, e, mis_bloques4)
print(mis_bloques4_encriptados)

[21101450083, 11090237809, 7835290949, 13916509906, 5797007492, 14961567488, 18719985437, 11324383840, 18923852140, 1177562528, 2065877116, 12633916785, 12390310674, 7860888701, 13973961240, 21447248359, 6649317939, 1858340868, 9495270785, 9409415684, 528397340, 15319582691, 4613537384, 15302543808, 11322365081, 9863969465]


5) ¿Cuál es el mensaje oculto en la siguiente lista de enteros, si se sabe que se ha usado el RSA con p=104729, q=224737 y e=554039341?

[21820245060, 16973863343, 16647907597, 5351085797, 18648503124, 18273004418, 15817170431, 7381433519, 15585396102, 4443772741, 1656337393, 14754433457, 22946763748, 5135880178, 6542531552, 2051320888, 18050988245, 4994899630, 17668156117, 11269362579]


In [17]:
d = inverse_of(e, phi_m)
print(blocks2message(decode_rsa(m, d, mis_bloques3_encriptados)))

me opongo a toda supersticion sea musulmana sea cristiana sea judia o sea budist


6) ¿Cuál es el mensaje oculto en la siguiente lista de enteros, si se sabe que se ha usado el RSA con p=104729, q=224737 y e=554039341?

[20071683819, 18754823901, 15281925738, 22431087988, 18578136847, 14565167928, 20363208149, 23164154654, 19883031733, 17926748637, 22540995446, 9667978832, 3356165928, 20768696878, 7232548727, 17445960483, 17408203534, 18047748783, 11480314056, 20135029456, 21596106087, 21071138629, 12152023796, 18248646084, 9103249568, 22310161636]

In [18]:
d = inverse_of(e, phi_m)
print(blocks2message(decode_rsa(m, d, mis_bloques4_encriptados)))

no creas conveniente actuar ocultando pruebas pues las pruebas terminan por salir a la luz a vista de to


7. Usando RSA con p=163841, q=350377 y e=4180452401, encripte el siguiente mensaje:

`"paren el mundo que me quiero bajar"`

8. Usando RSA con p=163847, q=350381 y e=8584779379, encripte el siguiente mensaje:

`"estos son mis principios y si no le gustan le tengo otros"`

In [19]:
p = nt.prime(15001)
q = nt.prime(30001)
print("p: ", p)
print("q: ", q)
phi_m = (p-1) *(q-1)

p:  163847
q:  350381


In [20]:
e = random_coprime_with(phi_m)
e

28480286763

Para manuscrito

In [21]:
m = 319
e = 3
phi_m = 280
mensaje5 = "saludos a todos"


In [22]:
mis_bloques5 = message2blocks(mensaje5, possible_len=2)
mis_bloques5

[30, 11, 22, 32, 14, 26, 30, 38, 11, 38, 31, 26, 14, 26, 30]

In [23]:
mis_bloques5_encriptados = encode_rsa(m, e, mis_bloques5)

In [24]:
mis_bloques5_encriptados

[204, 55, 121, 230, 192, 31, 204, 4, 55, 4, 124, 31, 192, 31, 204]

In [25]:
d = inverse_of(e, phi_m)
print(blocks2message(decode_rsa(m, d, mis_bloques5_encriptados)))

saludos a todos


In [31]:
p = 23
q = 29
m = p * q
e = 5
men = "duerme bien"
blo = message2blocks(men, possible_len=2)

enc = encode_rsa(m, e, blo)
print("encriptado")
print(enc)
d = inverse_of(e, 22*28)


encriptado
[222, 330, 329, 232, 460, 329, 237, 41, 195, 329, 645]


Para encriptar un mensaje, el primer paso es convertirlo en una secuencia de números como lo indica la siguiente tabla:
|A|B|C|D|E|F|G|H|I|J|K|L|M|N|Ñ|O|P|Q|R|S|T|U|V|W|X|Y|Z|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
|11|12|13|14|15|16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31|32|33|34|35|36|37|

y al espacio entre palabras le asignamos los dígitos $38$. Así, "hola mundo" se convierte en 18262211382332241426.

|H|O|L|A| |M|U|N|D|O|
|---|---|---|---|---|---|---|---|---|---|
|18|26|22|11|38|23|32|24|14|26|

Para realizar en manuscrito.

1) Usando el módulo $m=pq=319$ y el exponente $e=3$, que es coprimo con $\varphi(m) = 280$, encripte el mensaje `saludos a todos`.

2) Hace mucho, mucho tiempo en una galaxia muy lejana, te encontrabas en una misión ultrasecreta llevando en tu memoria las claves secretas $p=23$, $q=29$ y exponente $e=5$. Llegas al Hotel Electiva2, donde te alojas, para "descansar" después de un día arduo de trabajo y estudio; poco después de la media noche alguien toca a tu puerta y un misterioso mensajero te deja un papel con el mensaje encriptado siguiente:  

[222, 330, 329, 232, 460, 329, 237, 41, 195, 329, 645]

Tu misión, si la aceptas, es aprovar el primer parcial del Hotel Elevtiva2, si sigues la orden del mensaje anterior.